app/v3/asset/internal/keeper/ante.go

```diff
@@ -17,64 +16,38 @@ func NewAnteHandler(k Keeper) sdk.AnteHandler {
 	) (newCtx sdk.Context, res sdk.Result, abort bool) {
 		// new ctx
 		newCtx = sdk.Context{}
-
-		// get the signing accouts
-		signerAccs := auth.GetSigners(ctx)
-		if len(signerAccs) == 0 {
-			return newCtx, types.ErrSignersMissingInContext(types.DefaultCodespace, "signers missing in context").Result(), true
-		}
-
-		// get the payer
-		payer := signerAccs[0]
-
 		// total fee
-		totalFee := sdk.Coins{}
-
+		feeMap := make(map[string]sdk.Coins)
 		for _, msg := range tx.GetMsgs() {
 			// only check consecutive msgs which are routed to asset from the beginning
 			if msg.Route() != types.MsgRoute {
 				break
 			}
+			sender := msg.GetSigners()[0].String()
+			if _, ok := feeMap[sender]; !ok {
+				feeMap[sender] = sdk.Coins{}
+			}
 
-			var msgFee sdk.Coin
-
+			var fee sdk.Coin
 			switch msg := msg.(type) {
-			case types.MsgCreateGateway:
-				msgFee = GetGatewayCreateFee(ctx, k, msg.Moniker)
-				break
-
 			case types.MsgIssueToken:
-				if msg.Source == types.NATIVE {
-					msgFee = GetTokenIssueFee(ctx, k, msg.Symbol)
-				} else if msg.Source == types.GATEWAY {
-					msgFee = GetGatewayTokenIssueFee(ctx, k, msg.Symbol)
-				}
-
+				fee = k.getTokenIssueFee(ctx, msg.Symbol)
 				break
-
 			case types.MsgMintToken:
-				prefix, symbol := types.GetTokenIDParts(msg.TokenId)
-
-				if prefix == "" || prefix == "i" {
-					msgFee = GetTokenMintFee(ctx, k, symbol)
-				} else if prefix != "x" {
-					msgFee = GetGatewayTokenMintFee(ctx, k, symbol)
-				}
-
+				_, symbol := types.GetTokenIDParts(msg.TokenId)
+				fee = k.getTokenMintFee(ctx, symbol)
 				break
-
-			default:
-				msgFee = sdk.NewCoin(sdk.IrisAtto, sdk.ZeroInt())
 			}
-
-			totalFee = totalFee.Add(sdk.Coins{msgFee})
+			feeMap[sender] = feeMap[sender].Add(sdk.NewCoins(fee))
 		}
 
-		if !totalFee.IsAllLTE(payer.GetCoins()) {
-			// return error result and abort
-			return newCtx, types.ErrInsufficientCoins(types.DefaultCodespace, fmt.Sprintf("insufficient coins for asset fee: %s needed", totalFee.MainUnitString())).Result(), true
+		for addr, fee := range feeMap {
+			owner, _ := sdk.AccAddressFromBech32(addr)
+			balance := k.bk.GetCoins(ctx, owner)
+			if balance.IsAllLT(fee) {
+				return newCtx, types.ErrInsufficientCoins(types.DefaultCodespace, fmt.Sprintf("insufficient coins for asset fee: %s needed", fee.MainUnitString())).Result(), true
+			}
```

Comment:
On `DeliverTx`, the nondeterminacy of map traversal will cause different `gas_used`.